File: LLM-Translations/GPT-4/codenet/Go/Java/s301190573.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class s301190573 {
    static int size = 1000000;
    static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer tokenizer;

    public static void main(String[] args) throws IOException {
        int n, m;
        String[] input = nextLine().split(" ");
        n = Integer.parseInt(input[0]);
        m = Integer.parseInt(input[1]);
        int[] p = nextIntArray();
        int x, y;
        initialize(n);
        for (int i = 0; i < m; i++) {
            input = nextLine().split(" ");
            x = Integer.parseInt(input[0]) - 1;
            y = Integer.parseInt(input[1]) - 1;
            unite(x, y);
        }
        int s = 0;
        for (int i = 0; i < n; i++) {
            if (same(i, p[i])) {
                s++;
            }
        }
        System.out.println(s);
    }

    static int[] par, rank;

    static void initialize(int n) {
        par = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) {
            par[i] = i;
        }
    }

    static int root(int i) {
        if (par[i] == i) {
            return i;
        }
        par[i] = root(par[i]);
        return par[i];
    }

    static boolean same(int i, int j) {
        return root(i) == root(j);
    }

    static void unite(int i, int j) {
        i = root(i);
        j = root(j);
        if (i == j) {
            return;
        }
        if (rank[i] < rank[j]) {
            par[i] = j;
        } else {
            par[j] = i;
            if (rank[i] == rank[j]) {
                rank[i]++;
            }
        }
    }

    static String nextLine() throws IOException {
        return reader.readLine();
    }

    static int[] nextIntArray() throws IOException {
        tokenizer = new StringTokenizer(reader.readLine());
        int[] a = new int[tokenizer.countTokens()];
        for (int i = 0; i < a.length; i++) {
            a[i] = Integer.parseInt(tokenizer.nextToken()) - 1;
        }
        return a;
    }
}

// 

File: LLM-Translations/GPT-4/codenet/Go/Python/s301190573.py
import sys
from typing import List, Tuple

sys.setrecursionlimit(1000000)

def next_line() -> str:
    return input().strip()

def next_ints() -> Tuple[int, int]:
    x, y = map(int, next_line().split())
    return x, y

def next_int_array() -> List[int]:
    a = list(map(int, next_line().split()))
    return [i - 1 for i in a]

par = []
rank = []

def main():
    global par, rank
    n, m = next_ints()
    p = next_int_array()

    initialize(n)
    
    for _ in range(m):
        x, y = next_ints()
        unite(x - 1, y - 1)

    s = 0
    for i in range(n):
        if same(i, p[i]):
            s += 1

    print(s)

def initialize(n: int):
    global par, rank
    par = [i for i in range(n)]
    rank = [0] * n

def root(i: int) -> int:
    global par
    if par[i] == i:
        return i
    par[i] = root(par[i])
    return par[i]

def same(i: int, j: int) -> bool:
    return root(i) == root(j)

def unite(i: int, j: int):
    global par, rank
    i = root(i)
    j = root(j)
    if i == j:
        return
    if rank[i] < rank[j]:
        par[i] = j
    else:
        par[j] = i
        if rank[i] == rank[j]:
            rank[i] += 1

if __name__ == "__main__":
    main()


